File: veriblock-core/src/main/java/org/veriblock/core/Debug.java
package org.veriblock.core;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.veriblock.core.wallet.AddressManager;

import java.io.BufferedReader;
import java.io.File;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.io.StringWriter;

public class Debug {
    private static final Logger logger = LoggerFactory.getLogger(Debug.class);
    public static void main(String[] args) throws IOException {
        String installedPackages = getInstalledPackagesUbuntu();
        logger.info("Installed packages (Ubuntu):");
        logger.info(installedPackages);

        AddressManager manager = new AddressManager();
        manager.load(new File("wallet.dat"));
        for (int i = 0; i < 100_000; i++) {
            int entropyBefore = Integer.parseInt(getEntropy().replaceAll("\"", "").trim());
            long start = System.currentTimeMillis();
            String address = manager.getNewAddress().getHash();
            int entropyAfter = Integer.parseInt(getEntropy().replaceAll("\"", "").trim());
            logger.info("Generating address #" + i + " took " + (System.currentTimeMillis() - start) + "ms. Entropy before: " + entropyBefore + ", entropy after: " + entropyAfter);
        }
    }

    public static String getEntropy() {
        String line = null;
        String strstatus = "";
        try {

            String[] cmd = {"/bin/sh", "-c", "cat /proc/sys/kernel/random/entropy_avail"};
            Process p = Runtime.getRuntime().exec(cmd);
            BufferedReader in = new BufferedReader(new InputStreamReader(p.getInputStream()));
            while ((line = in.readLine()) != null) {
                strstatus += line + "\n";
            }
            in.close();
        } catch (Exception e) {

            StringWriter sw = new StringWriter();
            PrintWriter pw = new PrintWriter(sw);
            e.printStackTrace(pw);
            pw.flush();
            String stackTrace = sw.toString();
            int lenoferrorstr = stackTrace.length();
            if (lenoferrorstr > 500) {
                strstatus = "Error:" + stackTrace.substring(0, 500);
            } else {
                strstatus = "Error:" + stackTrace.substring(0, lenoferrorstr - 1);
            }
        }
        return strstatus;
    }

    public static String getInstalledPackagesUbuntu() {
        String line = null;
        String strstatus = "";
        try {

            String[] cmd = {"/bin/sh", "-c", "dpkg -l"};
            Process p = Runtime.getRuntime().exec(cmd);
            BufferedReader in = new BufferedReader(new InputStreamReader(p.getInputStream()));
            while ((line = in.readLine()) != null) {
                strstatus += line + "\n";
            }
            in.close();
        } catch (Exception e) {

            StringWriter sw = new StringWriter();
            PrintWriter pw = new PrintWriter(sw);
            e.printStackTrace(pw);
            pw.flush();
            String stackTrace = sw.toString();
            int lenoferrorstr = stackTrace.length();
            if (lenoferrorstr > 500) {
                strstatus = "Error:" + stackTrace.substring(0, 500);
            } else {
                strstatus = "Error:" + stackTrace.substring(0, lenoferrorstr - 1);
            }
        }
        return strstatus;
    }
}

File: veriblock-core/debug/src/main/kotlin/org/veriblock/core/debug/Debug.kt

@file:JvmName("Debug")

package org.veriblock.core.debug

import io.ktor.client.HttpClient
import io.ktor.client.engine.cio.CIO
import io.ktor.client.request.get
import io.ktor.client.statement.HttpResponse
import io.ktor.utils.io.jvm.javaio.copyTo
import org.veriblock.core.utilities.createLogger
import org.veriblock.core.wallet.AddressManager
import java.io.File
import java.io.PrintWriter
import java.io.StringWriter
import java.nio.file.Paths
import kotlin.system.exitProcess
import kotlin.system.measureTimeMillis

val logger = createLogger {}

private val httpClient = HttpClient(CIO)

suspend fun main(args: Array<String>) {
    if (args.isEmpty()) {
        logger.error { "The program requires to be passed a wallet file URL as an argument" }
        exitProcess(1)
    }

    val fileUrl = args[0]

    printDiagnostics()

    val downloadedWalletFile = File("downloaded_wallet.dat")
    downloadedWalletFile.delete()
    try {
        httpClient.get<HttpResponse>(fileUrl).content.copyTo(
            downloadedWalletFile.outputStream()
        )
    } catch (e: Exception) {
        logger.error { "Unable to download wallet file $fileUrl" }
        exitProcess(1)
    }

    logger.info { "Testing tx creation on default wallet..." }
    testWalletFile(File("wallet.dat"))

    logger.info { "Testing tx creation on downloaded wallet..." }
    testWalletFile(downloadedWalletFile)

    logger.info { "Tests finished! You can find the logs at ${File(".").canonicalPath}/debug.log" }
}

private fun testWalletFile(file: File) {
    val manager = AddressManager().apply {
        load(file)
    }
    repeat(10) {
        val entropyBefore = getEntropy().replace("\"", "").trim().toInt()
        val timeTaken = measureTimeMillis { manager.getNewAddress() }
        val entropyAfter = getEntropy().replace("\"", "").trim().toInt()
        logger.info { "Generating address #$it took ${timeTaken}ms. Entropy before: $entropyBefore, entropy after: $entropyAfter" }
    }
}

fun getEntropy() = getCommandResult("cat /proc/sys/kernel/random/entropy_avail")

fun getInstalledPackagesUbuntu() = getCommandResult("dpkg -l")

fun getCommandResult(command: String): String = try {
    val cmd = arrayOf("/bin/sh", "-c", command)
    val p = Runtime.getRuntime().exec(cmd)
    p.inputStream.readBytes().decodeToString()
} catch (e: Exception) {
    val sw = StringWriter()
    val pw = PrintWriter(sw)
    e.printStackTrace(pw)
    pw.flush()
    val stackTrace = sw.toString()
    stackTrace.substring(0, (stackTrace.length - 1).coerceAtMost(500))
}
